CMakeLists.txt

```diff
@@ -63,14 +63,37 @@ if (NOT ZLIB_FOUND)
     endif()
 endif()
 
-find_package(GMP)
-if (NOT GMP_FOUND)
-    if (APPLE)
-        message(FATAL_ERROR "GMP not found. Try 'brew install gmp'.")
-    elseif (UNIX)
-        message(FATAL_ERROR "GMP not found. Try 'sudo apt install libgmp-dev' or 'sudo dnf install gmp-devel'.")
-    else()
-        message(FATAL_ERROR "GMP not found.")
+option(DOWNLOAD_GMP "Download libgmp and build the library locally instead of using a system installation" OFF)
+if (DOWNLOAD_GMP)
+    include(ExternalProject)
+    ExternalProject_Add(
+        gmp
+        URL https://gmplib.org/download/gmp/gmp-6.3.0.tar.xz
+        INSTALL_DIR ${CMAKE_CURRENT_BINARY_DIR}/gmp
+        CONFIGURE_COMMAND "<SOURCE_DIR>/configure" "--prefix=<INSTALL_DIR>" --disable-shared --enable-static
+        USES_TERMINAL_CONFIGURE TRUE
+        BUILD_IN_SOURCE FALSE
+        USES_TERMINAL_BUILD TRUE
+        USES_TERMINAL_INSTALL TRUE
+    )
+    ExternalProject_Get_property(gmp INSTALL_DIR)
+    add_library(GMP::GMP STATIC IMPORTED GLOBAL)
+    # Work around `Imported target "GMP::GMP" includes non-existent path`:
```

Comment:
How come you get this error?

Reply:
I am not sure why this happens but it seems to only check presence of the include dir during CMake time (even if I add a BYPRODUCTS, to indicate that it will be created during the build). We might be able to turn it into a warning instead with: https://cmake.org/cmake/help/latest/policy/CMP0027.html (https://gitlab.kitware.com/cmake/cmake/-/blob/v3.27.4/Source/cmGeneratorTarget.cxx#L3658)

Seems to be a known issue: https://stackoverflow.com/questions/45516209/cmake-how-to-use-interface-include-directories-with-externalproject

Reply:
Ah ok maybe we should add a comment linking to https://gitlab.kitware.com/cmake/cmake/-/issues/15052